sklearn/model_selection/split.py

```diff
@@ -0,0 +1,1111 @@
+            n_iter, test_size, train_size, random_state)
+
+    def _iter_indices(self, X, y, labels=None):
+        n = _num_samples(X)
+        n_train, n_test = _validate_shuffle_split(n, self.test_size,
+                                                  self.train_size)
+        classes, y_indices = np.unique(y, return_inverse=True)
+        n_cls = classes.shape[0]
+
+        if np.min(bincount(y_indices)) < 2:
+            raise ValueError("The least populated class in y has only 1"
+                             " member, which is too few. The minimum"
+                             " number of labels for any class cannot"
+                             " be less than 2.")
+
+        if n_train < n_cls:
+            raise ValueError('The train_size = %d should be greater or '
+                             'equal to the number of classes = %d' %
+                             (n_train, n_cls))
+        if n_test < n_cls:
+            raise ValueError('The test_size = %d should be greater or '
+                             'equal to the number of classes = %d' %
+                             (n_test, n_cls))
+
+        rng = check_random_state(self.random_state)
+        cls_count = bincount(y_indices)
+        p_i = cls_count / float(n)
+        n_i = np.round(n_train * p_i).astype(int)
+        t_i = np.minimum(cls_count - n_i,
+                         np.round(n_test * p_i).astype(int))
+
+        for n in range(self.n_iter):
+            train = []
+            test = []
+
+            for i, cls in enumerate(classes):
+                permutation = rng.permutation(cls_count[i])
+                cls_i = np.where((y == cls))[0][permutation]
+
+                train.extend(cls_i[:n_i[i]])
+                test.extend(cls_i[n_i[i]:n_i[i] + t_i[i]])
+
+            # Because of rounding issues (as n_train and n_test are not
+            # dividers of the number of elements per class), we may end
+            # up here with less samples in train and test than asked for.
+            if len(train) < n_train or len(test) < n_test:
+                # We complete by affecting randomly the missing indexes
+                missing_idx = np.where(bincount(train + test,
+                                                minlength=len(y)) == 0)[0]
+                missing_idx = rng.permutation(missing_idx)
+                train.extend(missing_idx[:(n_train - len(train))])
+                test.extend(missing_idx[-(n_test - len(test)):])
+
+            train = rng.permutation(train)
+            test = rng.permutation(test)
+
+            yield train, test
+
+    def n_splits(self, X=None, y=None, labels=None):
+        return self.n_iter
+
+
+class PredefinedSplit(_PartitionIterator):
+    """Predefined split cross validation iterator
+
+    Splits the data into training/test set folds according to a predefined
+    scheme. Each sample can be assigned to at most one test set fold, as
+    specified by the user through the ``test_fold`` parameter.
```

Comment:
docstring is inaccurate, and I think this is really stretching the meaning of `labels`.  Is there any setting when a predefined split isn't known at init time? I'm not quite sure what the right api is here.